aeon/schema/social.py

```diff
@@ -0,0 +1,88 @@
+"""Readers for data relevant to Social experiments."""
+
+import json
+from pathlib import Path
+
+import pandas as pd
+
+from aeon import util
+import aeon.io.reader as _reader
+
+
+class Pose(_reader.Harp):
+    """Reader for Harp-binarized tracking data given a model that outputs id, parts, and likelihoods."""
+
+    def __init__(self, pattern):
+        self.pattern = pattern
+        self.extension = "bin"
+
+    def read(self, file, ceph_proc_dir="/ceph/aeon/aeon/data/processed"):
+        """Reads data from the Harp-binarized tracking file."""
+        # Get config file from `file`, then bodyparts from config file.
+        model_dir = file.stem.replace("_", "/")
+        config_file_dir = Path(ceph_proc_dir + "/" + model_dir)
+        assert config_file_dir.exists(), f"Cannot find model dir {config_file_dir}"
+        config_file = get_config_file(config_file_dir)
+        parts = self.get_bodyparts(config_file)
+        # Using bodyparts, assign column names to Harp register values, and read data in default format.
+        columns = ["class", "class_likelihood"]
+        for part in parts:
+            columns.extend([f"{part}_x", f"{part}_y", f"{part}_likelihood"])
+        self.columns = columns
+        data = super().read(file)
+        # Set new columns, and reformat `data`.
+        new_columns = ["class", "class_likelihood", "part", "part_likelihood", "x", "y"]
+        new_data = pd.DataFrame(columns=new_columns)
+        part_data_list = [None] * len(parts)
+        for i, part in enumerate(parts):
+            part_columns = ["class", "class_likelihood", f"{part}_x", f"{part}_y", f"{part}_likelihood"]
+            part_data_list[i] = data[part_columns]
+            part_data_list[i].insert(2, "part", part)
+            part_data_list[i].columns = new_columns
+        new_data = pd.concat(part_data_list)
+        return new_data.sort_index()
```

Comment:
Why is the sort necessary? Can't we assume that parts are inserted in the other they arrive in the `bin` file? It shouldn't really matter anyway.

Reply:
The way parts are added, you add all the timestamps for one part, then all the timestamps for next part, etc. 

So the frame is sorted by parts, not by time, without this `sort_index()`

Reply:
I see, that makes sense but I still believe there must be a more efficient way to write this without the need for sorting.

The reason is that if you look at a sequence of parts in a harp message, it will have the shape: [p1t0x, p1t0y, p2t0x, p2t0y; p1t1x, p1t1y, p2t1x, p2t1y, ...]

If this was numpy, it would be possible to use reshape directly to get instead:

[p1t0x,p1t0y;
p2t0x,p2t0y;
p1t1x,p1t1y;
p2t1x,p1t1y]

This would avoid the need to sort at all. We might need to reshape before converting to pandas depending on how internally they represent columns. There are tricky bits with the repeating part names etc so needs some thought but I think it will be worth it.

Reply:
I actually think there may not be a better way to do this, under the assumption we're operating from the Super's (Harp's') `read` method. One obvious naive way to do it without sorting would be to iterate through the rows in `data` and assign them one-by-one to rows in `new_data`, but ofc this is more inefficient.

I agree sorting could be avoided, but I think it may be easiest if this was done on the Bonsai/Harp side, such that harp messagaes from Sleap inference take the final form we want here `["class", "class_likelihood", "part", "part_likelihood", "x", "y"]` instead of the current form `["class", "class_likelihood", "part_x", "part_y", "part_likelihood"]`. For this I think we'd just need to add a `part` int id to the message - can we do this @glopesdev ?

Reply:
Actually just thought of one other thing I could try:

generate timestamps as indices, use this to index into `data`, assign to `new_data` for each part and `n_parts` - 1 rows.

I think it'll make sense when I show the code. Will try this and show the code and profile and lyk

Reply:
Sounds good, lmk how it goes. If necessary we can also easily split just the raw numpy loading part of the super() into a separate method, so that you would have more flexibility of manipulation before everything gets turned into a dataframe.

Reply:
@glopesdev 

This is another way I thought to do it without sorting, but it's actually ~50% longer than using the current `sort` procedure, even on large dataframes (I simulated 1 hour worth of frames at 50 fps)

```
n_parts = len(parts)
new_columns = ["class", "class_likelihood", "part", "part_likelihood", "x", "y"]
new_data = pd.DataFrame(columns=new_columns, index=np.repeat(data.index, n_parts))
for i, part in enumerate(parts):
    part_columns = ["class", "class_likelihood", f"{part}_x", f"{part}_y", f"{part}_likelihood"]
    part_data = data[part_columns]
    part_data.insert(2, "part", part)
    part_data.columns = new_columns
    new_data.iloc[i::n_parts] = part_data
```

I guess the indexing in `new_data.iloc[i::n_parts]` in python is much slower than the compiled C of `sort_index`

Reply:
I would still like to take a jab at improving this implementation without sorting, but I'm happy to merge this as a first draft and I can open an issue / branch later for improving this.